internal/graph/lookup.go

```diff
@@ -43,864 +38,97 @@ type ValidatedLookupRequest struct {
-		request = cl.processRewrite(ctx, req, nsdef, typeSystem, rewrite)
-	} else {
-		request = cl.lookupDirect(ctx, req, typeSystem)
-	}
-
-	// Perform the structural lookup.
-	result := lookupAny(ctx, req, req.Limit, []ReduceableLookupFunc{request})
-	if result.Err != nil {
-		return returnResult(lookupResultError(req, result.Err, result.Resp.Metadata))
-	}
-
-	objSet.Update(result.Resp.ResolvedOnrs)
-
-	// Recursively perform lookup on any of the ONRs found that do not match the target ONR.
-	// This ensures that we resolve the full transitive closure of all objects.
-	toCheck := objSet
-	responseMetadata := result.Resp.Metadata
-	for toCheck.Length() != 0 && objSet.Length() < req.Limit {
-		var requests []ReduceableLookupFunc
-		for _, obj := range toCheck.AsSlice() {
-			// If we've already found the target ONR, no further resolution is necessary.
-			if onrEqualOrWildcard(obj, req.Subject) {
-				continue
-			}
-
-			requests = append(requests, cl.dispatch(ValidatedLookupRequest{
-				&v1.DispatchLookupRequest{
-					Subject:        obj,
-					ObjectRelation: req.ObjectRelation,
-					Limit:          req.Limit - objSet.Length(),
-					Metadata:       decrementDepth(req.Metadata),
-					DirectStack:    req.DirectStack,
-					TtuStack:       req.TtuStack,
-				},
-				req.Revision,
-			}))
-		}
-
-		result := lookupAny(ctx, req, req.Limit, requests)
-		responseMetadata = combineResponseMetadata(responseMetadata, result.Resp.Metadata)
-
-		if result.Err != nil {
-			return returnResult(lookupResultError(req, result.Err, responseMetadata))
-		}
-
-		toCheck = tuple.NewONRSet()
-
-		for _, obj := range result.Resp.ResolvedOnrs {
-			// Only check recursively for new objects.
-			if objSet.Add(obj) {
-				toCheck.Add(obj)
-			}
-		}
-	}
-
-	return returnResult(lookupResult(req, limitedSlice(objSet.AsSlice(), req.Limit), responseMetadata))
+	return nil
 }
 
-func (cl *ConcurrentLookup) lookupDirect(ctx context.Context, req ValidatedLookupRequest, typeSystem *namespace.TypeSystem) ReduceableLookupFunc {
-	requests := []ReduceableLookupFunc{}
-
-	// Ensure type informatione exists on the relation.
-	if !typeSystem.HasTypeInformation(req.ObjectRelation.Relation) {
-		return returnResult(lookupResultError(
-			req,
-			NewRelationMissingTypeInfoErr(req.ObjectRelation.Namespace, req.ObjectRelation.Relation),
-			emptyMetadata,
-		))
-	}
-
-	// Dispatch a check for the target ONR directly, if it is allowed on the start relation.
-	isDirectAllowed, err := typeSystem.IsAllowedDirectRelation(req.ObjectRelation.Relation, req.Subject.Namespace, req.Subject.Relation)
-	if err != nil {
-		return returnResult(lookupResultError(req, err, emptyMetadata))
+func (cl *ConcurrentLookup) LookupViaReachability(ctx context.Context, req ValidatedLookupRequest) (*v1.DispatchLookupResponse, error) {
+	if req.Subject.ObjectId == tuple.PublicWildcard {
+		resp := lookupResultError(req, NewErrInvalidArgument(errors.New("cannot perform lookup on wildcard")), emptyMetadata)
```

Comment:
If we made this a streaming API w/ paging, maybe this is not so unreasonable?

Reply:
Maybe, but I think for now its fine to keep this as-is